### PR TITLE
Make custom ops differentiable

### PR DESCRIPTION
**Pull request by t-vi**
Make custom ops differentiable and replace autograd.Function. Use ops unconditionally.

We may consider removing the extension module in a follow-up.
The code-path is tested by the existing tests for differentiability.

### CODE REVIEW

File: torchvision/csrc/custom_ops/custom_ops.cpp [hunk 1]
```diff
@@ -6,9 +6,139 @@
...
+      Variable rois,
+      const double spatial_scale,
+      const int64_t pooled_height,
+      const int64_t pooled_width,
+      const int64_t sampling_ratio) {
+    ctx->saved_data["spatial_scale"] = spatial_scale;
+    ctx->saved_data["pooled_height"] = pooled_height;
+    ctx->saved_data["pooled_width"] = pooled_width;
+    ctx->saved_data["sampling_ratio"] = sampling_ratio;
+    ctx->saved_data["batch_size"] = input.size(0);
```
Comment (fmassa) [inline torchvision/csrc/custom_ops/custom_ops.cpp]: nit: This is fine as is, this is more a question for myself: I thought `IValue` allowed us to to also have lists of ints internally, is that right?

**Comment by fmassa** [comment]
@t-vi tests failures seems to be related
```
 /tmp/pip-req-build-4yu7qgl5/torchvision/csrc/custom_ops/custom_ops.cpp: In function ‘std::tuple<at::Tensor, at::Tensor> roi_pool(const at::Tensor&, const at::Tensor&, double, int64_t, int64_t)’:
  /tmp/pip-req-build-4yu7qgl5/torchvision/csrc/custom_ops/custom_ops.cpp:136:31: error: converting to ‘std::tuple<at::Tensor, at::Tensor>’ from initializer list would use explicit constructor ‘constexpr std::tuple<_T1, _T2>::tuple(_U1&&, _U2&&) [with _U1 = torch::autograd::Variable&; _U2 = torch::autograd::Variable&; <template-parameter-2-3> = void; _T1 = at::Tensor; _T2 = at::Tensor]’
     return {result[0], result[1]};
```

**Comment by t-vi** [comment]
Which CI failure is that? I'm seeing a lot of `data_ptr`-things, but I cannot seem to find that error message... :confused:

File: torchvision/csrc/custom_ops/custom_ops.cpp (hunk 1 above)
Comment (t-vi) [reply torchvision/csrc/custom_ops/custom_ops.cpp]: Right now it was a 1-1 transcription of the Python Function. I'll check.
If we are going to keep the backward signature as is (as opposed to changing it to IntList), the benefit will be marginal, but given that we're providing an example, too...

**Comment by fmassa** [comment]
Here is one example https://travis-ci.org/pytorch/vision/jobs/582603380 , line 954

I've re-run some of the CircleCI failures, which I think were still picking the patch before my fix

File: torchvision/csrc/custom_ops/custom_ops.cpp (hunk 1 above)
Comment (t-vi) [reply torchvision/csrc/custom_ops/custom_ops.cpp]: Works well, I switched this to `input_shape`, but we might change the function signature when we remove the Python extension.

**Comment by fmassa** [comment]
@t-vi can you rebase your PR with current master? This looks pretty good!

**Pull request merged**